provider/core/src/resource.rs

```diff
@@ -172,15 +246,67 @@ impl ResourceKey {
     /// use icu_provider::prelude::*;
     ///
     /// let resc_key = icu_provider::hello_world::key::HELLO_WORLD_V1;
-    /// let components = resc_key.get_components();
+    /// let components: Vec<&str> = resc_key
+    ///     .iter_components()
+    ///     .collect();
     ///
     /// assert_eq!(
     ///     ["core", "helloworld@1"],
-    ///     components.iter().collect::<Vec<&str>>()[..]
+    ///     components[..]
     /// );
     /// ```
-    pub fn get_components(&self) -> ResourceKeyComponents {
-        self.into()
+    pub fn iter_components(&self) -> impl Iterator<Item = &str> {
+        self.get_path().split('/')
+    }
+
+    /// Gets the first path component of a [`ResourceKey`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    ///
+    /// let resc_key = icu_provider::hello_world::key::HELLO_WORLD_V1;
+    /// assert_eq!("core", resc_key.get_component_0());
+    /// ```
+    pub fn get_component_0(&self) -> &str {
+        // This cannot fail because of the preconditions on path (at least one '/')
+        self.iter_components().next().unwrap()
+    }
+
+    /// Gets the second path component of a [`ResourceKey`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    ///
+    /// let resc_key = icu_provider::hello_world::key::HELLO_WORLD_V1;
+    /// assert_eq!("helloworld@1", resc_key.get_component_1());
+    /// ```
+    pub fn get_component_1(&self) -> &str {
+        // This cannot fail because of the preconditions on path (at least one '/')
+        self.iter_components().nth(1).unwrap()
+    }
+
+    /// Gets the last path component of a [`ResourceKey`] without the version suffix.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_provider::prelude::*;
+    ///
+    /// let resc_key = icu_provider::hello_world::key::HELLO_WORLD_V1;
+    /// assert_eq!("helloworld", resc_key.get_last_component_no_version());
+    /// ```
+    pub fn get_last_component_no_version(&self) -> &str {
```

Comment:
I don't like these names at all...

Also, you just told me that you prefer the constructor to take a string instead of (category, subcategory, version) so there's more flexibility in how many namespaces a client uses, but this API very rigidly only supports the latter.

Reply:
I deleted `get_component_0` and `get_component_1`.  I need `get_last_component_no_version` for compatibility with uprops provider.  I filed #1515 to follow up on that dependency.